### ISSUE TITLE
Migration workflow

### ISSUE THREAD

**Issue by romansavrulin** [DmitryTsepelev/store_model#32]
How can we handle data migration in underlying jsonb field?

Here's an example. I've defined the following structure for AssetIssueList v1

```ruby
class AssetIssue
  include StoreModel::Model

  attribute :key, :string, default: ''
  attribute :name, :string, default: ''
  attribute :exists, :boolean, default: false

  validates :key, name, :exists, presence: true
end

class AssetIssueList
  include StoreModel::Model

  attribute :version, :integer, default: 1
  attribute :list, AssetIssue.to_array_type, default: [ ]

  validates :version, :list, presence: true
end

class Asset < ApplicationRecord
  attribute :issues, AssetIssueList.to_type

  after_find :validateIssues
  def validateIssues
    if issues.nil? || issues.version != ISSUES_VERSION
      updateDefaultIssues
    end
  end
end
```

After some time, I've decided to remove `attribute :name` from AssetIssue. Application is unable to start due to error

```
ActiveModel::UnknownAttributeError (unknown attribute 'name' for AssetIssue.)

/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute_assignment.rb:53:in `_assign_attribute'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute_assignment.rb:44:in `block in _assign_attributes'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute_assignment.rb:43:in `each'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute_assignment.rb:43:in `_assign_attributes'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute_assignment.rb:35:in `assign_attributes'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/model.rb:81:in `initialize'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attributes.rb:80:in `initialize'
/usr/local/bundle/gems/store_model-0.5.2/lib/store_model/types/array_type.rb:78:in `new'
/usr/local/bundle/gems/store_model-0.5.2/lib/store_model/types/array_type.rb:78:in `block in ensure_model_class'
/usr/local/bundle/gems/store_model-0.5.2/lib/store_model/types/array_type.rb:77:in `map'
/usr/local/bundle/gems/store_model-0.5.2/lib/store_model/types/array_type.rb:77:in `ensure_model_class'
/usr/local/bundle/gems/store_model-0.5.2/lib/store_model/types/array_type.rb:34:in `cast_value'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/type/value.rb:38:in `cast'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute.rb:174:in `type_cast'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute.rb:42:in `value'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute_set.rb:41:in `fetch_value'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attributes.rb:130:in `attribute'
/usr/local/bundle/gems/activemodel-6.0.0/lib/active_model/attribute_methods.rb:383:in `list'

/app/models/navigation_asset.rb:14:in `validateIssues'
```

How to deal with that? Couldn't be not existent static keys (name in my example) just dynamic in  `AssetIssueList` so we can migrate underlying data storage with ease?

**Comment by romansavrulin** [comment]
I've digged into the code and found that `json_type` has rescue

https://github.com/DmitryTsepelev/store_model/blob/2066b1ea31639473ac652a114dfda2080eb77c5a/lib/store_model/types/json_type.rb#L69-L76

But `array_type` seems not

https://github.com/DmitryTsepelev/store_model/blob/2066b1ea31639473ac652a114dfda2080eb77c5a/lib/store_model/types/array_type.rb#L76-L80

**Comment by DmitryTsepelev** [comment]
Hi @romansavrulin!

That's a fair point, array elements should be handled in the exact same way. I'd really appreciate a PR or I'll implement it by myself a little bit later 🙂

As an alternative solution, the same result could be reached by adding a regular SQL migration that will update the structure of the JSON (I guess it would be helpful to document this process)

**Comment by romansavrulin** [comment]
@DmitryTsepelev I have never developed or fixed a ruby gem before, but I can try. Can you provide a link to any good starting point guide about how to install this gem as a git submodule to an exiting project, instead of using a bundler definition?

**Comment by DmitryTsepelev** [comment]
The easiest way is to change the gem itself:

1. Fork the gem and clone it to your machine
2. Navigate to the gem folder and run `bundle install`
3. Run specs (it's just a regular `rspec`, specs should pass)
4. Add new specs for `ArrayType` (similar to [these specs](https://github.com/DmitryTsepelev/store_model/blob/master/spec/store_model/types/json_type_spec.rb#L67))
5. Fix the code to pass the specs
6. Commit and make and PR, CI will make sure specs are passing

If you want to play with your own version of the gem as a part of a real project, you can open up project's `Gemfile` and replace `gem "store_model"` with `gem "store_model", path: "/path/to/the/gem/on/your/local/machine"` and run `bundle install` – it will use your local version. 

Please let me know if you have any questions or issues 🙂

**Comment by alexeevit** [comment]
What behavior is expected?

- `ArrayType#unkown_attributes` returns hash contains all unknown attributes from all the elements
or
- Each of the array elements has it's own `#unknown_attributes`

The first one is little strange because it will merge all `#unkown_attributes` hashes and replace elements which has equal keys. But the description of the task on cultofmartians is contains exactly this behavior.

**Comment by DmitryTsepelev** [comment]
@alexeevit the second one, we need to have an array of objects with `#unknown_attributes ` filled for them (I'll update the description at cult since it sounds confusing 🙂)

**Comment by DmitryTsepelev** [comment]
Released a fix as part of `0.5.3`. Closing the issue for now, please let me know if something goes wrong
